### ISSUE TITLE
Pre-calculate required data size for messages before serializing

### ISSUE THREAD

**Issue by allevato** [apple/swift-protobuf#34]
Using the new performance harness, I tested serialization of a message with 100 fixed32 fields. 80% of the harness's time was spent in the `serializeProtobuf()` method, with nearly 97% of the time in *that* method spent in `RangeReplaceableCollection.append()` (pardon the poor indentation; thanks Instruments):

```
Weight	Self Weight		Symbol Name
239.00 ms  100.0%	0 s	 	Harness.(run() -> ()).(closure #1)
191.00 ms   79.9%	0 s	 	 ProtobufBinaryMessageBase.serializeProtobuf() throws -> Data
185.00 ms   77.4%	3.00 ms	 	  RangeReplaceableCollection.append<A where ...> (contentsOf : A1) -> ()
4.00 ms    1.6%	0 s	 	  protocol witness for ProtobufBinaryMessageBase.serializeProtobuf() throws -> Data in conformance PerfMessage
2.00 ms    0.8%	2.00 ms	 	  small_free_list_add_ptr
46.00 ms   19.2%	0 s	 	 ProtobufMessage.init(protobuf : Data) throws -> A
1.00 ms    0.4%	1.00 ms	 	 PerfMessage.field97.setter
1.00 ms    0.4%	0 s	 	 swift_deallocClassInstance
```

In other words, most of the time is being spent appending new content to the buffer as it's being serialized. We can gain a **huge** performance advantage if we pre-calculate the required buffer size instead and then pass an already-allocated buffer into the encoder.

I'll start working on a branch to test this. It'll be interesting to see how much of a boost this gives us, because it gives us a better basis for comparing binary serialization using the visitor approach against something more explicit and sequential.

**Comment by tbkka** [comment]
The trivial test would just pre-allocate 16k buffers always and see what happens then.  In particular, this would help identify whether the time is being spent resizing the underlying store or somewhere else.

**Comment by tbkka** [comment]
A related question:  Is it faster to append a few bytes at a time to a Data() than to a [UInt8]?

**Comment by jcanizales** [comment]
I don't see it mentioned explicitly in the docs, but it looks like appending Data to Data might not copy the bytes (so O(1) instead of O(n))?

**Comment by allevato** [comment]
Fixed by #43.
